pkg/proctree/export.go

```diff
@@ -0,0 +1,214 @@
+package proctree
+
+/*
+Package proctree is used to create and maintain a process tree using Tracee's events.
+
+The process tree exports to the user only read operations - the user can query for a processes and
+threads information.
+However, the process tree itself is rather complicated, and shouldn't be changed by unsupported
+methods.
+Currently, the only supported way to update and change the process tree is via events. The process
+tree is aware of Tracee's events, and know how to update itself using events. It can be done
+using a pipeline or directly using the appropriate method.
+
+# Synchronization
+
+The tree usage is not synchronized.
+Process information might be requested some time after it was updated, or even deleted.
+To address this issue, the tree:
+*	Makes processes available even after they are exited and should be deleted for some time.
+*	Exports processes and threads information according to a specific time, instead of giving the
+	entire information and leaving the user to process it.
+
+# Events feeding edge cases handling
+
+The overall design of the tree is taking into consideration that events are not guaranteed to be
+ordered, nor they are guaranteed to arrive at all.
+Moreover, the tree needs to take into account that it starts running in the middle of the system
+runtime. As a result, it missed some processes fork, exec and exit events.
+To withstand this issues, the tree do the following things:
+*	Limit the amount of nodes it keeps, to avoid leakage when missing exit events.
+*	Take information from all events, even those which are not related to process and thread life
+	cycle.
+*	Try cleaning processes with PID which is found to be reused by new fork events.
+*	Initializing parent processes nodes, even with almost no information, to establish connections
+	between known nodes.
+*	The tree is *not* trying to create one big tree, but instead has the forest architecture.
+	As it doesn't have all the history, it can't connect all nodes, and so it connects only nodes
+	it knows are connected.
+*/
```

Comment:
```suggestion
// Process Tree Documentation
//
// The process tree only exports read operations to the user - the user can query for processes and
// threads information. However, the process tree itself is rather complex and should not be changed
// by unsupported methods.
//
// Currently, the only supported way to update and change the process tree is via events. The
// process tree is aware of Tracee's events and knows how to update itself using these events. This
// can be done either using a pipeline or directly using the appropriate method.
//
// # Synchronization
//
// The tree's usage is not synchronized. Process information may be requested some time after it has
// been updated or even deleted. To address this issue, the tree:
//
// 1. Makes processes available even after they have exited and should have been deleted for some
//    time.
// 2. Exports processes and threads information according to a specific time, instead of providing
//    all the information and leaving the user to process it.
//
// # Events feeding and edge case handling
//
// The overall design of the tree takes into consideration that events are not guaranteed to be
// ordered, nor are they guaranteed to arrive at all. Moreover, the tree needs to account for the
// fact that it starts running in the middle of the system runtime. As a result, it misses some
// process fork, exec, and exit events.
//
// To withstand these issues, the tree does the following:
//
// 1. Limits the number of nodes it keeps to avoid leakage when missing exit events.
// 2. Gathers information from all events, even those that are not related to process and thread
//    life cycles.
// 3. Attempts to clean processes with a PID that is found to be reused by new fork events.
// 4. Initializes parent process nodes, even with minimal information, to establish connections
//    between known nodes.
// 5. The tree does *not* attempt to create one large tree, but instead adopts a forest
//    architecture. Since it does not have all the history, it can't connect all nodes, so it only
//    connects nodes it knows are connected.
//
```